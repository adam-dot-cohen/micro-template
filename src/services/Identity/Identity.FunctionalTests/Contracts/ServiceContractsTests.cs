using System;
using System.Linq;
using System.Reflection;
using Google.Protobuf.Reflection;
using Grpc.Core;
using Identity.Api.V1;
using Laso.Identity.Core.Extensions;
using Xunit;

namespace Laso.Identity.FunctionalTests.Contracts
{
    public class ServiceContractsTests
    {
        [Fact]
        public void Proto_files_should_not_have_breaking_changes()
        {
            typeof(Partners).Assembly.GetTypes()
                .Select(x => x.GetCustomAttribute<BindServiceMethodAttribute>(false)?.BindType)
                .Where(x => x != null)
                .Select(x => ((ServiceDescriptor) x.GetProperty(nameof(Partners.Descriptor)).GetValue(null)).File)
                .Distinct(x => x.Name)
                .ToList()
                .To(x => new
                {
                    Methods = x
                        .SelectMany(y => y.Services, (y, z) => new { Proto = y.Name, Service = z })
                        .SelectMany(y => y.Service.Methods, (y, z) => new
                        {
                            y.Proto,
                            Service = y.Service.FullName,
                            z.Name,
                            InputType = z.InputType.FullName,
                            OutputType = z.OutputType.FullName
                        }),
                    Fields = x
                        .SelectMany(y => y.MessageTypes, (y, z) => new { Proto = y.Name, MessageType = z })
                        .SelectMany(y => y.MessageType.Fields.InFieldNumberOrder(), (y, z) => new
                        {
                            y.Proto,
                            MessageType = y.MessageType.FullName,
                            z.Name,
                            z.Index,
                            Type = GetType(z),
                            z.IsRequired,
                            z.IsRepeated
                        }),
                    EnumValues = x
                        .SelectMany(y => y.EnumTypes, (y, z) => new { Proto = y.Name, EnumType = z })
                        .SelectMany(y => y.EnumType.Values, (y, z) => new
                        {
                            y.Proto,
                            EnumType = y.EnumType.FullName,
                            z.Name,
                            z.Number
                        })
                })
                .ApproveIf(x =>
                {
                    return x.Approved.Methods.All(y => x.Received.Methods.Contains(y))
                           && x.Approved.Fields.All(y => x.Received.Fields.Contains(y))
                           && x.Approved.EnumValues.All(y => x.Received.EnumValues.Contains(y));
                });
        }

        private static string GetType(FieldDescriptor field)
        {
            switch (field.FieldType)
            {
                case FieldType.Message:
                    return field.MessageType.FullName;
                case FieldType.Enum:
                    return field.EnumType.FullName;
                case FieldType.Group:
                    throw new NotSupportedException();
                default:
                    return field.FieldType.ToString();
            }
        }
    }
}
