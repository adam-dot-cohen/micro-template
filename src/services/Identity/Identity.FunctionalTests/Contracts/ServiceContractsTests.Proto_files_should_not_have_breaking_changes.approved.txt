{
  "Methods": [
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "Service": "identity.api.v1.Partners",
      "Name": "CreatePartner",
      "InputType": "identity.api.v1.CreatePartnerRequest",
      "OutputType": "identity.api.v1.CreatePartnerReply"
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "Service": "identity.api.v1.Partners",
      "Name": "GetPartners",
      "InputType": "identity.api.v1.GetPartnersRequest",
      "OutputType": "identity.api.v1.GetPartnersReply"
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "Service": "identity.api.v1.Partners",
      "Name": "GetPartner",
      "InputType": "identity.api.v1.GetPartnerRequest",
      "OutputType": "identity.api.v1.GetPartnerReply"
    }
  ],
  "Fields": [
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.CreatePartnerRequest",
      "Name": "partner",
      "Type": "identity.api.v1.PartnerModel",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.CreatePartnerReply",
      "Name": "id",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerModel",
      "Name": "id",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerModel",
      "Name": "name",
      "Type": "String",
      "FieldNumber": 2,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerModel",
      "Name": "contactName",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 3,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerModel",
      "Name": "contactPhone",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 4,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerModel",
      "Name": "contactEmail",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 5,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerModel",
      "Name": "publicKey",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 6,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerModel",
      "Name": "normalizedName",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 7,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerView",
      "Name": "id",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerView",
      "Name": "name",
      "Type": "String",
      "FieldNumber": 2,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerView",
      "Name": "contactName",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 3,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerView",
      "Name": "contactPhone",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 4,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerView",
      "Name": "contactEmail",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 5,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.GetPartnersReply",
      "Name": "partners",
      "Type": "identity.api.v1.PartnerView",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": true,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.GetPartnerRequest",
      "Name": "id",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.GetPartnerReply",
      "Name": "partner",
      "Type": "identity.api.v1.PartnerView",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    }
  ],
  "EnumValues": []
}